### DISCUSSION TITLE
Testing: In memory data doesn't persist per test case

### DISCUSSION THREAD

**Discussion by michaelxswe** [fastapi/fastapi#11358] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel
from sqlalchemy.orm import Session
from fastapi import Depends, FastAPI, HTTPException


class UserCreateModel(BaseModel):
    username: str
    password: str


def get_database_session():
    yield


class UserRepo:
    def create_user(self, user_create_model: UserCreateModel, session: Session):
        pass

    def get_user_by_id(self, id: int, session: Session):
        pass


def create_app():
    app = FastAPI()

    @app.post("/api/v1/users", status_code=201)
    async def create_user(
        user_create_model: UserCreateModel, user_repo: UserRepo = Depends(), session=Depends(get_database_session)
    ):
        user = user_repo.create_user(user_create_model=user_create_model, session=session)
        return user

    @app.get("/api/v1/users/{id}", status_code=200)
    async def get_user_by_id(id: int, user_repo: UserRepo = Depends(), session=Depends(get_database_session)):
        user = user_repo.get_user_by_id(id=id, session=session)

        if not user:
            raise HTTPException(status_code=404)
        return user

    return app


app = create_app()
```

### Description

Below is my test set up

```
from app import UserCreateModel, UserRepo, create_app, get_database_session
from dataclasses import dataclass
from fastapi import FastAPI
from fastapi.testclient import TestClient
from pytest import fixture
from sqlalchemy.orm import Session


@fixture(scope="session")
def app():
    def _get_database_session():
        return True
    
    app = create_app()
    app.dependency_overrides[get_database_session] = _get_database_session
    yield app


@fixture(scope="session")
def client(app: FastAPI):
    client = TestClient(app=app)
    yield client


@fixture(scope="function")
def user_repo(app: FastAPI):
    print("created")

    @dataclass
    class User:
        id: int
        username: str
        password: str

    class MockUserRepo:
        def __init__(self):
            self.database = []

        def create_user(self, user_create_model: UserCreateModel, session: Session) -> User | None:
            if session:
                user = User(
                    id=len(self.database) + 1,
                    username=user_create_model.username,
                    password=user_create_model.password,
                )
                self.database.append(user)
                return user

        def get_user_by_id(self, id: int, session: Session) -> User | None:
            print(self.database)
            if session:
                for user in self.database:
                    if user.id == id:
                        return user

                return None


    app.dependency_overrides[UserRepo] = MockUserRepo
    yield


def test_create_user(client: TestClient, user_repo):
    data = {"username": "mike", "password": "123"}
    res = client.post("/api/v1/users", json=data)

    assert res.status_code == 201
    assert res.json()["username"] == "mike"
    assert res.json()["password"] == "123"
    assert "id" in res.json()

def test_get_user(client: TestClient, user_repo):
    data = {"username": "nick", "password": "123"}
    res = client.post("/api/v1/users", json=data)
    assert res.status_code == 201
    user = res.json()
    user_id = user["id"]

    res = client.get(f"/api/v1/users/{user_id}")

    assert res.status_code == 200

```

For test_get_user(), I first make a post request and append {"username": "nick", "password": "123"} to self.database, then I follow up with a get request, and yet my self.database is still empty even though the fixture is set to function scope. Does anyone know what is happening?

### Operating System

Windows

### FastAPI Version

latest

### Pydantic Version

latest

### Python Version

Python 3.11.1

**Comment by YuriiMotov** [comment, marked as answer]
That's because in your second test, `res = client.post("/api/v1/users", json=data)` and `res = client.get(f"/api/v1/users/{user_id}")` are two different API calls and FastAPI will resolve dependencies for each of them by evoking `UserRepo()` and this will create new `MockUserRepo` instance every time.

To solve you problem you need to modify you test code so that your dependency always returns the same `UserRepo` instance during the one test.
In your `user_repo` fixture change
```
    app.dependency_overrides[UserRepo] = MockUserRepo
    yield
```
to 
```
    repo = MockUserRepo()

    app.dependency_overrides[UserRepo] = lambda : repo
    yield
```
Now, your overrided `UserRepo` dependency will always return the same `repo` object during one test. And your test works.

**Reply by michaelxswe** [reply]
Thanks!

**Reply by michaelxswe** [reply]
Just to confirm my understanding, with your solution, the the state of dependency persist throughout the whole test, and then it's reset when it moves to the next test case?

**Reply by YuriiMotov** [reply]
Yes, that's because the scope of your `user_repo` fixture is 'function'
